File: android/src/main/java/com/margelo/nitro/sharedtransition/HybridSharedTransitionModule.kt
package com.margelo.nitro.sharedtransition

import android.app.Activity
import android.app.Application
import android.graphics.Bitmap
import android.graphics.Canvas
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.lang.ref.WeakReference

/**
 * Native module for shared element transitions on Android
 *
 * Provides Fabric-safe APIs for:
 * - View snapshot capture using View.draw()
 * - Precise layout measurement in screen coordinates
 * - Element visibility control
 *
 * Uses modern Android APIs:
 * - View.draw() for snapshot capture
 * - No deprecated UIManager/findNodeHandle APIs
 */
@DoNotStrip
@Keep
class HybridSharedTransitionModule : HybridSharedTransitionModuleSpec() {

    // Cached snapshots by nativeID (URI -> file path)
    private val cachedSnapshots = mutableMapOf<String, String>()

    // Hidden elements by nativeID
    private val hiddenElements = mutableSetOf<String>()

    // Clone views by view tag
    private val cloneViews = mutableMapOf<Int, View>()

    // Counter for clone view tags
    private var cloneViewTagCounter = 1000

    // Main thread handler
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun measureNode(nativeId: String): Promise<SharedElementNodeData> {
        return Promise.async {
            withContext(Dispatchers.Main) {
                val view = findViewByNativeID(nativeId)
                    ?: throw Exception("View not found: $nativeId")

                val location = IntArray(2)
                view.getLocationOnScreen(location)

                val contentType = detectContentType(view)
                val snapshotUri = captureSnapshotSync(view, nativeId)

                SharedElementNodeData(
                    layout = SharedElementLayout(
                        x = location[0].toDouble(),
                        y = location[1].toDouble(),
                        width = view.width.toDouble(),
                        height = view.height.toDouble()
                    ),
                    contentType = contentType,
                    snapshotUri = snapshotUri
                )
            }
        }
    }

    override fun captureSnapshot(nativeId: String): Promise<String> {
        return Promise.async {
            withContext(Dispatchers.Main) {
                val view = findViewByNativeID(nativeId)
                    ?: throw Exception("View not found: $nativeId")
                captureSnapshotSync(view, nativeId)
            }
        }
    }

    override fun prepareTransition(
        startNodeId: String,
        endNodeId: String,
        config: TransitionConfig
    ): Promise<PreparedTransitionData> {
        return Promise.async {
            withContext(Dispatchers.Main) {
                val startView = findViewByNativeID(startNodeId)
                    ?: throw Exception("Start view not found: $startNodeId")

                val endView = findViewByNativeID(endNodeId)
                    ?: throw Exception("End view not found: $endNodeId")

                // Get screen locations
                val startLocation = IntArray(2)
                val endLocation = IntArray(2)
                startView.getLocationOnScreen(startLocation)
                endView.getLocationOnScreen(endLocation)

                // Capture snapshots
                val startSnapshotUri = captureSnapshotSync(startView, startNodeId)
                val endSnapshotUri = captureSnapshotSync(endView, endNodeId)

                // Detect content types
                val startContentType = detectContentType(startView)
                val endContentType = detectContentType(endView)

                val startLayout = SharedElementLayout(
                    x = startLocation[0].toDouble(),
                    y = startLocation[1].toDouble(),
                    width = startView.width.toDouble(),
                    height = startView.height.toDouble()
                )

                val endLayout = SharedElementLayout(
                    x = endLocation[0].toDouble(),
                    y = endLocation[1].toDouble(),
                    width = endView.width.toDouble(),
                    height = endView.height.toDouble()
                )

                PreparedTransitionData(
                    startLayout = startLayout,
                    endLayout = endLayout,
                    startSnapshotUri = startSnapshotUri,
                    endSnapshotUri = endSnapshotUri,
                    startContentType = startContentType,
                    endContentType = endContentType
                )
            }
        }
    }

    override fun createCloneView(nativeId: String): Promise<Double> {
        return Promise.async {
            withContext(Dispatchers.Main) {
                val originalView = findViewByNativeID(nativeId)
                    ?: throw Exception("View not found: $nativeId")

                val activity = ActivityHolder.currentActivity
                    ?: throw Exception("No activity available")

                // Create bitmap of the view
                val bitmap = Bitmap.createBitmap(
                    originalView.width,
                    originalView.height,
                    Bitmap.Config.ARGB_8888
                )
                val canvas = Canvas(bitmap)
                originalView.draw(canvas)

                // Create ImageView with the bitmap
                val cloneView = ImageView(activity).apply {
                    setImageBitmap(bitmap)
                    scaleType = ImageView.ScaleType.FIT_XY
                }

                // Position in screen coordinates
                val location = IntArray(2)
                originalView.getLocationOnScreen(location)

                // Get decoration view for overlay
                val decorView = activity.window.decorView as ViewGroup

                // Add to decoration view
                val layoutParams = ViewGroup.LayoutParams(
                    originalView.width,
                    originalView.height
                )
                decorView.addView(cloneView, layoutParams)

                // Position the clone
                cloneView.x = location[0].toFloat()
                cloneView.y = location[1].toFloat()

                // Generate tag
                val viewTag = cloneViewTagCounter++
                cloneViews[viewTag] = cloneView

                viewTag.toDouble()
            }
        }
    }

    override fun destroyCloneView(viewTag: Double) {
        val tag = viewTag.toInt()

        mainHandler.post {
            cloneViews[tag]?.let { cloneView ->
                (cloneView.parent as? ViewGroup)?.removeView(cloneView)
                cloneViews.remove(tag)
            }
        }
    }

    override fun setNodeHidden(nativeId: String, hidden: Boolean) {
        mainHandler.post {
            findViewByNativeID(nativeId)?.let { view ->
                view.visibility = if (hidden) View.INVISIBLE else View.VISIBLE

                if (hidden) {
                    hiddenElements.add(nativeId)
                } else {
                    hiddenElements.remove(nativeId)
                }
            }
        }
    }

    override fun cleanup() {
        mainHandler.post {
            // Remove all cached snapshots
            cachedSnapshots.values.forEach { uri ->
                if (uri.startsWith("file://")) {
                    File(uri.removePrefix("file://")).delete()
                }
            }
            cachedSnapshots.clear()

            // Show all hidden elements
            hiddenElements.forEach { nativeId ->
                findViewByNativeID(nativeId)?.visibility = View.VISIBLE
            }
            hiddenElements.clear()

            // Remove all clone views
            cloneViews.values.forEach { view ->
                (view.parent as? ViewGroup)?.removeView(view)
            }
            cloneViews.clear()
        }
    }

    // =========================================================================
    // Private Helpers
    // =========================================================================

    /**
     * Capture snapshot synchronously (must be called on main thread)
     */
    private fun captureSnapshotSync(view: View, nativeId: String): String {
        // Check cache
        cachedSnapshots[nativeId]?.let { return it }

        // Ensure view has valid size
        if (view.width <= 0 || view.height <= 0) {
            throw Exception("View has zero size")
        }

        // Create bitmap
        val bitmap = Bitmap.createBitmap(
            view.width,
            view.height,
            Bitmap.Config.ARGB_8888
        )

        // Draw view to canvas
        val canvas = Canvas(bitmap)
        view.draw(canvas)

        // Get cache directory from activity
        val activity = ActivityHolder.currentActivity
            ?: throw Exception("No activity for cache directory")

        // Save to file
        val safeNativeId = nativeId.replace("/", "_").replace(":", "_").replace(".", "_")
        val fileName = "shared_transition_${safeNativeId}_${System.currentTimeMillis()}.png"
        val file = File(activity.cacheDir, fileName)

        FileOutputStream(file).use { out ->
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
        }

        val uri = "file://${file.absolutePath}"
        cachedSnapshots[nativeId] = uri

        return uri
    }

    /**
     * Detect the content type of a view
     */
    private fun detectContentType(view: View): SharedElementContentType {
        // Check if it's an ImageView
        if (view is ImageView) {
            return SharedElementContentType.IMAGE
        }

        // Check if it contains an ImageView as direct child
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                if (view.getChildAt(i) is ImageView) {
                    return SharedElementContentType.IMAGE
                }
            }
        }

        // Default to snapshot
        return SharedElementContentType.SNAPSHOT
    }

    /**
     * Find a view by its nativeID prop (Fabric-safe)
     */
    private fun findViewByNativeID(nativeId: String): View? {
        val activity = ActivityHolder.currentActivity ?: return null
        val rootView = activity.window.decorView.rootView as? ViewGroup ?: return null
        return findViewRecursive(rootView, nativeId)
    }

    /**
     * Recursive view search with multiple strategies for Fabric compatibility
     */
    private fun findViewRecursive(viewGroup: ViewGroup, nativeId: String): View? {
        for (i in 0 until viewGroup.childCount) {
            val child = viewGroup.getChildAt(i)

            // Strategy 1: Check content description (from accessibilityLabel)
            if (child.contentDescription == nativeId) {
                return child
            }

            // Strategy 2: Check tag as String
            if (child.tag == nativeId) {
                return child
            }

            // Strategy 3: Try React Native's getNativeId() method
            try {
                val method = child.javaClass.getMethod("getNativeId")
                if (method.invoke(child) == nativeId) return child
            } catch (_: Exception) {}

            // Strategy 4: Try mNativeId field (ReactViewGroup)
            try {
                val field = child.javaClass.getDeclaredField("mNativeId")
                field.isAccessible = true
                if (field.get(child) == nativeId) return child
            } catch (_: Exception) {}

            // Recurse into children
            if (child is ViewGroup) {
                findViewRecursive(child, nativeId)?.let { return it }
            }
        }
        return null
    }

    companion object {
        @Suppress("unused")
        private const val TAG = "SharedTransitionModule"
    }
}

/**
 * Static holder for the current Activity.
 * Used by Nitro Modules to access the activity without ReactApplicationContext.
 */
@DoNotStrip
@Keep
object ActivityHolder {
    private var activityRef: WeakReference<Activity>? = null
    private var isInitialized = false

    val currentActivity: Activity?
        get() = activityRef?.get()

    /**
     * Initialize the activity lifecycle callbacks
     */
    fun init(application: Application) {
        if (isInitialized) return
        isInitialized = true

        application.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
                activityRef = WeakReference(activity)
            }

            override fun onActivityStarted(activity: Activity) {
                activityRef = WeakReference(activity)
            }

            override fun onActivityResumed(activity: Activity) {
                activityRef = WeakReference(activity)
            }

            override fun onActivityPaused(activity: Activity) {}
            override fun onActivityStopped(activity: Activity) {}
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
            override fun onActivityDestroyed(activity: Activity) {
                if (activityRef?.get() === activity) {
                    activityRef = null
                }
            }
        })
    }

    /**
     * Manually set the current activity (called from React Native)
     */
    fun setCurrentActivity(activity: Activity?) {
        activityRef = activity?.let { WeakReference(it) }
    }
}

File: ios/HybridSharedTransitionModule.swift
import UIKit
import Foundation
import NitroModules

/**
 * Native module for shared element transitions on iOS
 *
 * Provides Fabric-safe APIs for:
 * - View snapshot capture using CALayer rendering
 * - Precise layout measurement in screen coordinates
 * - Element visibility control
 *
 * Uses modern iOS APIs (iOS 13+):
 * - UIGraphicsImageRenderer for efficient rendering
 * - UIWindowScene for window access
 * - No deprecated UIManager/findNodeHandle APIs
 */
class HybridSharedTransitionModule: HybridSharedTransitionModuleSpec {

  // MARK: - Properties

    /// Cached snapshots by nativeID (URI -> file path)
    private var cachedSnapshots: [String: URL] = [:]

    /// Hidden elements by nativeID
    private var hiddenElements: Set<String> = []

    /// Clone views by view tag
    private var cloneViews: [Int: UIView] = [:]

    /// Counter for clone view tags
    private var cloneViewTagCounter: Int = 1000

  // MARK: - HybridSharedTransitionModuleSpec Implementation

    func measureNode(nativeId: String) throws -> Promise<SharedElementNodeData> {
        return Promise.async {
        guard let view = self.findViewByNativeID(nativeId) else {
                throw NSError(
              domain: "SharedTransition",
              code: 404,
              userInfo: [NSLocalizedDescriptionKey: "View not found: \(nativeId)"]
            )
        }

            // Get frame in window coordinates (screen-relative)
            guard let window = view.window else {
                throw NSError(
                    domain: "SharedTransition",
                    code: 500,
                    userInfo: [NSLocalizedDescriptionKey: "View not in window hierarchy: \(nativeId)"]
                )
            }

            let frameInWindow = view.convert(view.bounds, to: window)

            // Detect content type
            let contentType = self.detectContentType(view)

            // Capture snapshot
            let snapshotUri = try self.captureSnapshotSync(view: view, nativeId: nativeId)

            let layout = SharedElementLayout(
                x: Double(frameInWindow.origin.x),
                y: Double(frameInWindow.origin.y),
                width: Double(frameInWindow.width),
                height: Double(frameInWindow.height)
          )

            return SharedElementNodeData(
                layout: layout,
                contentType: contentType,
                snapshotUri: snapshotUri
            )
    }
  }

    func captureSnapshot(nativeId: String) throws -> Promise<String> {
        return Promise.async {
        guard let view = self.findViewByNativeID(nativeId) else {
                throw NSError(
              domain: "SharedTransition",
              code: 404,
              userInfo: [NSLocalizedDescriptionKey: "View not found: \(nativeId)"]
            )
            }

            return try self.captureSnapshotSync(view: view, nativeId: nativeId)
        }
    }

    func prepareTransition(
        startNodeId: String,
        endNodeId: String,
        config: TransitionConfig
    ) throws -> Promise<PreparedTransitionData> {
        return Promise.async {
            // Find both views
            guard let startView = self.findViewByNativeID(startNodeId) else {
                throw NSError(
                    domain: "SharedTransition",
                    code: 404,
                    userInfo: [NSLocalizedDescriptionKey: "Start view not found: \(startNodeId)"]
                )
            }

            guard let endView = self.findViewByNativeID(endNodeId) else {
                throw NSError(
                    domain: "SharedTransition",
                    code: 404,
                    userInfo: [NSLocalizedDescriptionKey: "End view not found: \(endNodeId)"]
                )
            }

            // Get window for coordinate conversion
            guard let window = startView.window else {
                throw NSError(
              domain: "SharedTransition",
              code: 500,
                    userInfo: [NSLocalizedDescriptionKey: "Views not in window hierarchy"]
            )
            }

            // Measure layouts
            let startFrame = startView.convert(startView.bounds, to: window)
            let endFrame = endView.convert(endView.bounds, to: window)

            // Capture snapshots
            let startSnapshotUri = try self.captureSnapshotSync(view: startView, nativeId: startNodeId)
            let endSnapshotUri = try self.captureSnapshotSync(view: endView, nativeId: endNodeId)

            // Detect content types
            let startContentType = self.detectContentType(startView)
            let endContentType = self.detectContentType(endView)

            let startLayout = SharedElementLayout(
                x: Double(startFrame.origin.x),
                y: Double(startFrame.origin.y),
                width: Double(startFrame.width),
                height: Double(startFrame.height)
            )

            let endLayout = SharedElementLayout(
                x: Double(endFrame.origin.x),
                y: Double(endFrame.origin.y),
                width: Double(endFrame.width),
                height: Double(endFrame.height)
            )

            return PreparedTransitionData(
                startLayout: startLayout,
                endLayout: endLayout,
                startSnapshotUri: startSnapshotUri,
                endSnapshotUri: endSnapshotUri,
                startContentType: startContentType,
                endContentType: endContentType
            )
        }
  }

    func createCloneView(nativeId: String) throws -> Promise<Double> {
        return Promise.async {
            guard let originalView = self.findViewByNativeID(nativeId) else {
                throw NSError(
                    domain: "SharedTransition",
                    code: 404,
                    userInfo: [NSLocalizedDescriptionKey: "View not found: \(nativeId)"]
                )
            }

            guard let window = originalView.window else {
                throw NSError(
                    domain: "SharedTransition",
                    code: 500,
                    userInfo: [NSLocalizedDescriptionKey: "View not in window"]
                )
            }

            // Create snapshot view
            let snapshotView = originalView.snapshotView(afterScreenUpdates: false)
                ?? UIView(frame: originalView.bounds)

            // Position in window coordinates
            let frameInWindow = originalView.convert(originalView.bounds, to: window)
            snapshotView.frame = frameInWindow

            // Generate tag
            let viewTag = self.cloneViewTagCounter
            self.cloneViewTagCounter += 1

            // Store and add to window
            self.cloneViews[viewTag] = snapshotView
            window.addSubview(snapshotView)

            return Double(viewTag)
        }
    }

    func destroyCloneView(viewTag: Double) throws {
        let tag = Int(viewTag)

        if let cloneView = cloneViews[tag] {
            cloneView.removeFromSuperview()
            cloneViews.removeValue(forKey: tag)
        }
    }

    func setNodeHidden(nativeId: String, hidden: Bool) throws {
        guard let view = findViewByNativeID(nativeId) else {
            return // Silently ignore if view not found
        }

        DispatchQueue.main.async {
            view.isHidden = hidden

            if hidden {
                self.hiddenElements.insert(nativeId)
            } else {
                self.hiddenElements.remove(nativeId)
      }
    }
  }

  func cleanup() throws {
        // Remove all cached snapshots
        for (_, url) in cachedSnapshots {
        try? FileManager.default.removeItem(at: url)
      }
    cachedSnapshots.removeAll()

        // Show all hidden elements
        for nativeId in hiddenElements {
            if let view = findViewByNativeID(nativeId) {
                view.isHidden = false
            }
        }
        hiddenElements.removeAll()

        // Remove all clone views
        for (_, view) in cloneViews {
            view.removeFromSuperview()
        }
        cloneViews.removeAll()
  }

  // MARK: - Private Helpers

    /// Capture snapshot synchronously (must be called on main thread)
    private func captureSnapshotSync(view: UIView, nativeId: String) throws -> String {
        // Check if already cached
        if let cachedUrl = cachedSnapshots[nativeId] {
            return cachedUrl.absoluteString
        }

        // Ensure view has valid size
        guard view.bounds.width > 0 && view.bounds.height > 0 else {
            throw NSError(
                domain: "SharedTransition",
                code: 400,
                userInfo: [NSLocalizedDescriptionKey: "View has zero size"]
            )
        }

        // Use UIGraphicsImageRenderer for modern, efficient rendering
        let renderer = UIGraphicsImageRenderer(bounds: view.bounds)
        let image = renderer.image { context in
            // Use layer rendering for accurate capture including subviews
            view.layer.render(in: context.cgContext)
        }

        // Generate unique filename
        let safeNativeId = nativeId.replacingOccurrences(of: "/", with: "_")
        let fileName = "shared_transition_\(safeNativeId)_\(Date().timeIntervalSince1970).png"
        let fileURL = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)

        // Save as PNG (lossless, supports transparency)
        guard let pngData = image.pngData() else {
            throw NSError(
                domain: "SharedTransition",
                code: 500,
                userInfo: [NSLocalizedDescriptionKey: "Failed to create PNG data"]
            )
        }

        try pngData.write(to: fileURL)

        // Cache the URL
        cachedSnapshots[nativeId] = fileURL

        return fileURL.absoluteString
    }

    /// Detect the content type of a view
    private func detectContentType(_ view: UIView) -> SharedElementContentType {
        // Check if it's an image view
        if view is UIImageView {
            return .image
        }

        // Check if it contains an image view as direct child
        for subview in view.subviews {
            if subview is UIImageView {
                return .image
            }
        }

        // Default to snapshot
        return .snapshot
    }

  /// Find a view by its nativeID prop (Fabric-safe)
  private func findViewByNativeID(_ nativeId: String) -> UIView? {
        // Get key window from connected scenes (iOS 13+)
        guard let windowScene = UIApplication.shared.connectedScenes
      .compactMap({ $0 as? UIWindowScene })
            .first(where: { $0.activationState == .foregroundActive }),
              let window = windowScene.windows.first(where: { $0.isKeyWindow }) else {
            // Fallback for older iOS or edge cases
            return findViewByNativeIDFallback(nativeId)
    }

    return findView(withNativeID: nativeId, in: window)
  }

    /// Fallback window access
    private func findViewByNativeIDFallback(_ nativeId: String) -> UIView? {
        guard let window = UIApplication.shared.windows.first(where: { $0.isKeyWindow }) else {
            return nil
        }
        return findView(withNativeID: nativeId, in: window)
    }

    /// Recursive view search
  private func findView(withNativeID nativeId: String, in view: UIView) -> UIView? {
        // Check accessibilityIdentifier (React Native maps nativeID to this)
    if view.accessibilityIdentifier == nativeId {
      return view
    }

        // Check React Native's nativeID property using KVC (safe)
        if let viewNativeID = (view as AnyObject).value(forKeyPath: "reactTag") as? String,
         viewNativeID == nativeId {
        return view
    }

        // Search subviews recursively
    for subview in view.subviews {
      if let found = findView(withNativeID: nativeId, in: subview) {
        return found
      }
    }

    return nil
  }
}
